Adding a --task option

snukky opened the pull request:
This PR will:
* add a `--problem` option (mentioned in #394) with example aliases for "transformer" and "transformer-big"
* add `--dump-config explain` for dumping a config file with expanded alias options
* bypass some validation rules if `--dump-config` is used, e.g. training sets and vocabs don't have to be set

**Note:** the pull request adds changes to the `some-updates` branch, not `master`. It is separated from PR #405 to have a clear view on modifications and a distinct space for discussion.

emjotde commented:
Did we say the option should be called "problem" in the end? It sounds so negative (might be honest though). I'd prefer "experiment" or maybe "task"?

emjotde commented:
Ha, I actually said "problem". Funny, I hate it now :)

emjotde commented:
God.... I think I am getting old and moody. I want "task" now.

snukky commented:
Renamed. I also like "task" much more than "problem". We should check if we can add  more aliases in this pull request and if the currently defined "transformer" and "transformer-big" are OK.

emjotde commented on src/common/cli_wrapper.cpp:
```diff
@@ -128,6 +128,38 @@ void CLIWrapper::parse(int argc, char **argv) {
   }
 }
 
+void CLIWrapper::parseAliases() {
```
Since this is a new functionality, can you add more comments about this function and also inside the function?

snukky replied:
Function documentation was in the header file. Added more comments inside the function.

emjotde commented on src/common/cli_wrapper.cpp:
```diff
@@ -128,6 +128,38 @@ void CLIWrapper::parse(int argc, char **argv) {
 }
 
+void CLIWrapper::parseAliases() {
+  if(aliases_.empty())
```
aliases are YAML nodes?

emjotde replied:
No, AliasTuple apparently, this deserves a comment, where to find the actual object definition.

emjotde commented on src/common/cli_wrapper.cpp:
```diff
@@ -128,6 +128,38 @@ void CLIWrapper::parse(int argc, char **argv) {
+
+      if(expand) {
+        updateConfig(alias.config,
+                     "Unknown option(s) in alias '" + alias.key + ": " + alias.value + "'");
```
Is this a hidden abort message? When does it fire?

snukky replied:
See documentation of updateConfig() in the header file.

emjotde commented on src/common/cli_wrapper.cpp:
```diff
@@ -128,6 +128,38 @@ void CLIWrapper::parse(int argc, char **argv) {
+    }
+  }
+
+  // Remove aliases from the config
```
Comment: Why?

emjotde commented on src/common/cli_wrapper.h:
```diff
@@ -51,16 +51,21 @@ class CLIFormatter : public CLI::Formatter {
   bool modified{false};
 };
 
+// Helper structure used for aliases and storing an option key, value, and YAML node
```
Maybe also explain what an alias is.

emjotde commented on src/common/config_parser.cpp:
```diff
@@ -79,7 +79,8 @@ void ConfigParser::addOptionsGeneral(cli::CLIWrapper& cli) {
   cli.add<std::string>("--dump-config",
-     "Dump current (modified) configuration to stdout and exit. Possible values: full, minimal")
+     "Dump current (modified) configuration to stdout and exit. "
+     "Possible values: full, minimal, explain")
```
Add comment: what do the values mean. I think we already saw that "minimal" is actually a bit counterintuitive.
Please explain "explain". Is that a good name?

emjotde commented on src/common/config_parser.cpp:
```diff
@@ -413,8 +414,12 @@ void ConfigParser::addOptionsTraining(cli::CLIWrapper& cli) {
   // add ULR settings
   addSuboptionsULR(cli);
+
+  cli.add<std::vector<std::string>>("--task",
```
Another proposition for the name was "--defaults-for-task". I am conflicted.

snukky replied:
My vote for `--task` as it's shorter.

emjotde commented on src/common/config_parser.cpp:
```diff
@@ -687,11 +692,10 @@ void ConfigParser::addSuboptionsULR(cli::CLIWrapper& cli) {
   // The order of aliases does matter as later options overwrite earlier
 
-  if(config_["best-deep"].as<bool>()) {
+  cli.alias("best-deep", "true", [](YAML::Node& config) {
```
I think these should be moved to a separate file, either by moving the function, or by using an `#include`.

```
void ConfigParser::addAliases(cli::CLIWrapper& cli) {
#include "common/aliases.inc"
}
```

@frankseide Is this ugly?

emjotde replied:
Or better have the entire function inside `common/aliases.inc`

emjotde replied:
Roman suggested to just put the function definition into a separate aliases.cpp which is of course much better. I have done something like that in the `bert` branch.

emjotde commented on src/common/config_parser.cpp:
```diff
@@ -701,11 +705,42 @@ void ConfigParser::expandAliases(cli::CLIWrapper& cli) {
-  }
+  });
+
+  cli.alias("task", "transformer", [](YAML::Node& config) {
```
I guess these should also have some comments, I guess this is something I can do later.

emjotde commented on src/common/config_validator.cpp:
```diff
@@ -10,7 +10,10 @@ bool ConfigValidator::has(const std::string& key) const {
-ConfigValidator::ConfigValidator(const YAML::Node& config) : config_(config) {}
+ConfigValidator::ConfigValidator(const YAML::Node& config)
+    : config_(config),
+      dump_(config["dump-config"] && !config["dump-config"].as<std::string>().empty()
```
What is this doing?

emjotde commented on src/common/config_validator.h:
```diff
@@ -5,11 +5,10 @@
 class ConfigValidator {
 private:
   const YAML::Node& config_;
+  bool dump_{false};
```
Maybe a longer more descriptive name?

emjotde commented:
Problem: when `task` is used with `dump-config` the options that were specified on the command line are not not identified anymore as options that should override the task options. When marian is then called with

```
marian --task transformer --dim-emb 256 --dump-config > test.yml
```

This will not still train with `dim-emb: 512` because task overrides `dim-emb` from the config:
```
marian --config test.yml 
```

snukky commented:
Recent changes:
* added comments
* fixed the mentioned problem by introducing priority levels for options
* moved alias definitions to a separate file

Additional regression tests are in https://github.com/marian-nmt/marian-regression-tests/tree/some-updates
